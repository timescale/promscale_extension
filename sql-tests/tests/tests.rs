use insta::assert_snapshot;
use std::{
    env,
    mem::MaybeUninit,
    sync::{Mutex, Once},
};
use test_common::*;
use test_generator::test_resources;

const TESTDATA: &'static str = concat!(env!("CARGO_MANIFEST_DIR"), "/testdata");

fn init_lock() -> &'static Mutex<()> {
    static mut SINGLETON: MaybeUninit<Mutex<()>> = MaybeUninit::uninit();
    static ONCE: Once = Once::new();

    unsafe {
        ONCE.call_once(|| {
            let singleton = Mutex::new(());
            SINGLETON.write(singleton);
        });

        SINGLETON.assume_init_ref()
    }
}

// TODO fix how test_resources works in nexted workspaces
#[test_resources("sql-tests/testdata/*.sql")]
fn sql_tests(full_resource: &str) {
    let pg_blueprint = PostgresContainerBlueprint::new().with_testdata(TESTDATA);
    let test_pg_instance = new_test_instance_from_env(&pg_blueprint);
    let mut init_conn = test_pg_instance.connect();
    // This lock prevents multiple callers from running
    // CREATE EXTENSION simultaneously. In case of
    // promscale_extension concurrent CREATE EXTENSION causes
    // issues when migration incremental/003-users.sql executes.
    {
        let _init_lock = init_lock().lock().unwrap();
        init_conn
            .simple_query("CREATE EXTENSION promscale;")
            .expect("Unable to create extension promscale.");
    }

    let resource = if let Some((_, rest)) = full_resource.split_once('/') {
        rest
    } else {
        full_resource
    };
    let query_result = test_pg_instance.exec_sql_script(resource);
    assert_snapshot!(resource, query_result);
}
