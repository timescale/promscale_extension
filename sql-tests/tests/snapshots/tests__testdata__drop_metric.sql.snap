---
source: sql-tests/tests/tests.rs
expression: query_result
---
 plan  
-------
 1..49
(1 row)

                  ok                   
---------------------------------------
 ok 1 - none of the chunks are deleted
(1 row)

                       ok                        
-------------------------------------------------
 ok 2 - none of the series should be removed yet
(1 row)

                           ok                            
---------------------------------------------------------
 ok 3 - none of the series should be marked for deletion
(1 row)

                      ok                      
----------------------------------------------
 ok 4 - none of the labels should deleted yet
(1 row)

                          asserts_before_deletion                           
----------------------------------------------------------------------------
 ok 5 - after the first timestamp: expired chunks are gone
 ok 6 - after the first timestamp: none of the series should be removed yet
 ok 7 - after the first timestamp: one series should be marked for deletion
 ok 8 - after the first timestamp: none of the labels should deleted yet
(4 rows)

                    asserts_before_deletion                     
----------------------------------------------------------------
 ok 9 - after iter 0: expired chunks are gone
 ok 10 - after iter 0: none of the series should be removed yet
 ok 11 - after iter 0: one series should be marked for deletion
 ok 12 - after iter 0: none of the labels should deleted yet
(4 rows)

                    asserts_before_deletion                     
----------------------------------------------------------------
 ok 13 - after iter 1: expired chunks are gone
 ok 14 - after iter 1: none of the series should be removed yet
 ok 15 - after iter 1: one series should be marked for deletion
 ok 16 - after iter 1: none of the labels should deleted yet
(4 rows)

                    asserts_before_deletion                     
----------------------------------------------------------------
 ok 17 - after iter 2: expired chunks are gone
 ok 18 - after iter 2: none of the series should be removed yet
 ok 19 - after iter 2: one series should be marked for deletion
 ok 20 - after iter 2: none of the labels should deleted yet
(4 rows)

                    asserts_before_deletion                     
----------------------------------------------------------------
 ok 21 - after iter 3: expired chunks are gone
 ok 22 - after iter 3: none of the series should be removed yet
 ok 23 - after iter 3: one series should be marked for deletion
 ok 24 - after iter 3: none of the labels should deleted yet
(4 rows)

                    asserts_before_deletion                     
----------------------------------------------------------------
 ok 25 - after iter 4: expired chunks are gone
 ok 26 - after iter 4: none of the series should be removed yet
 ok 27 - after iter 4: one series should be marked for deletion
 ok 28 - after iter 4: none of the labels should deleted yet
(4 rows)

                    asserts_after_deletion                     
---------------------------------------------------------------
 ok 29 - after iter 5: expired chunks are gone
 ok 30 - after iter 5: one series should be removed
 ok 31 - after iter 5: no series should be marked for deletion
 ok 32 - after iter 5: unused labels should deleted
(4 rows)

                    asserts_after_deletion                     
---------------------------------------------------------------
 ok 33 - after iter 6: expired chunks are gone
 ok 34 - after iter 6: one series should be removed
 ok 35 - after iter 6: no series should be marked for deletion
 ok 36 - after iter 6: unused labels should deleted
(4 rows)

                    asserts_after_deletion                     
---------------------------------------------------------------
 ok 37 - after iter 7: expired chunks are gone
 ok 38 - after iter 7: one series should be removed
 ok 39 - after iter 7: no series should be marked for deletion
 ok 40 - after iter 7: unused labels should deleted
(4 rows)

                    asserts_after_deletion                     
---------------------------------------------------------------
 ok 41 - after iter 8: expired chunks are gone
 ok 42 - after iter 8: one series should be removed
 ok 43 - after iter 8: no series should be marked for deletion
 ok 44 - after iter 8: unused labels should deleted
(4 rows)

                        asserts_after_deletion                         
-----------------------------------------------------------------------
 ok 45 - after all iterations: expired chunks are gone
 ok 46 - after all iterations: one series should be removed
 ok 47 - after all iterations: no series should be marked for deletion
 ok 48 - after all iterations: unused labels should deleted
(4 rows)

                     throws_like                      
------------------------------------------------------
 ok 49 - Epoch has changed after a series was dropped
(1 row)

 finish 
--------
(0 rows)


